File: mpi-1x-hypack-2013/matrix-vector-matrix-comp-codes/c-lang-matvect/vv_mult_blkstp_nonunf.c

/*
********************************************************************

		C-DAC Tech Workshop : hyPACK-2013
                           October 15-18, 2013

   Example 5.2		: vv_mult_blkstp_nonunf.c

   Objective           : Vector_Vector Multiplication 
                         (Using Non uniform Data Partitioning)

   Input               : Read files (vdata1.inp) for Vector_A and 
                         (vdata2.inp) for Vector_B 

   Output              : Result of Vector-Vector Multiplication on process 0

   Necessary Condition : Number of Processes should be
                         less than or equal to 8

   Created             : August-2013

   E-mail              : hpcfte@cdac.in     

******************************************************************/

#include <stdio.h>
#include <math.h>
#include <mpi.h>
#include <stdlib.h>

#define Epsilon 1.0E-10
main(int argc, char** argv) 
{

  int	   Numprocs, MyRank;
  int 	VectorSize, VectorSize_A, VectorSize_B; 
  int	   index, iproc, jproc, disp_value;
  int	   Root = 0, ValidInput = 1, ValidOutput = 1, DataSize;
  int    Distribute_Cols, Remaining_Cols, *SendCount, *Displacement;
  int    Destination, Destination_tag = 0, Source, Source_tag = 0;
  int    RecvCount;
  float  *Mybuffer_A, *Mybuffer_B,  MyFinalVector, FinalAnswer;
  float  CheckResultVector;
  FILE	*fp;
  int		VectorA_FileStatus = 1, VectorB_FileStatus = 1; 
  float  *Vector_A, *Vector_B;
  MPI_Status status;


/* ........MPI Initialisation .......*/

  MPI_Init(&argc, &argv); 
  MPI_Comm_rank(MPI_COMM_WORLD, &MyRank);
  MPI_Comm_size(MPI_COMM_WORLD, &Numprocs);

  if(MyRank == Root) {

/*.......Read the Vector A Input file ......*/
	  
     if((fp = fopen ("./data/vdata1.inp", "r")) == NULL) {
	     VectorA_FileStatus = 0;
  	  }

	  if(VectorA_FileStatus != 0) {
  	  	fscanf(fp, "%d\n", &VectorSize_A);     

/* .......Allocate memory and read data for vector A .....*/

          Vector_A  = (float *) malloc(VectorSize_A*sizeof(float));

/* .......Read data for matrix .....*/

     		for(index=0; index< VectorSize_A; index++)  {
     		  fscanf(fp, "%f", &Vector_A[index]); 
     		}
     		fclose(fp); 
     }

/*.......Read the Vector B Input file ......*/
	 
     if((fp = fopen ("./data/vdata2.inp", "r")) == NULL) {
	     VectorB_FileStatus = 0;
     }

	  if(VectorB_FileStatus != 0) {
     		fscanf(fp, "%d\n", &VectorSize_B);
	
/* .......Allocate memory and read data for vector B .....*/

            Vector_B = (float *)malloc(VectorSize_B*sizeof(float));
     	      for(index=0; index< VectorSize_B; index++) {
        	fscanf(fp, "%f", &Vector_B[index]);
     		}
    fclose(fp);
	    }

	  if(VectorSize_A != VectorSize_B)
	     ValidInput = 0;		  

	}
/* MyRank == 0 */

MPI_Barrier(MPI_COMM_WORLD);

MPI_Bcast(&VectorA_FileStatus, 1, MPI_INT, Root, MPI_COMM_WORLD);
 if(VectorA_FileStatus == 0) {
   if(MyRank == Root) printf("Can't open input file for Vector A\n");
   MPI_Finalize();
   exit(-1);
 }

MPI_Bcast(&VectorB_FileStatus, 1, MPI_INT, Root, MPI_COMM_WORLD);
   if(VectorB_FileStatus == 0) {
    if(MyRank == Root) printf("Can't open input file for Vector B\n");
    MPI_Finalize();
    exit(-1);
   }

MPI_Bcast(&ValidInput, 1, MPI_INT, Root, MPI_COMM_WORLD);

 if(ValidInput == 0) {
   if(MyRank == Root) printf("Size of both vectors is not equal .\n");
   MPI_Finalize();
   exit(-1);
   }

VectorSize = VectorSize_A;
MPI_Bcast(&VectorSize, 1, MPI_INT, Root, MPI_COMM_WORLD);

   if(VectorSize < Numprocs) {
     MPI_Finalize();
     if(MyRank == 0)
      printf("VectorSize should be more than No of Processors . \n");
     exit(0);
    }  	

/* Initial arrangement for Scatterv operation */

  if(MyRank == Root) {
     Displacement = (int *)malloc(Numprocs*sizeof(int));
     SendCount    = (int *)malloc(Numprocs*sizeof(int));

	Distribute_Cols = VectorSize/Numprocs;
	Remaining_Cols  = VectorSize % Numprocs;
	for(iproc = 0; iproc < Numprocs; iproc++)
          SendCount[iproc] = Distribute_Cols;

	for(iproc=Remaining_Cols; iproc>0; iproc--)
	  (SendCount[Remaining_Cols-iproc])++ ;

	Displacement[0] = 0;
	for(iproc = 1; iproc < Numprocs; iproc++) {
	  disp_value = 0;
	  for(jproc = 0; jproc < iproc; jproc++)
	     disp_value += SendCount[jproc];
	  Displacement[iproc] = disp_value;
        }

/* Send RecvCount to each process */

	for(iproc = 1; iproc < Numprocs; iproc++) {
	  Destination = iproc;
	  DataSize = 1;
	  RecvCount = SendCount[iproc];
          MPI_Send(&RecvCount, DataSize, MPI_INT, Destination, 
	  Destination_tag,MPI_COMM_WORLD);
        }
        RecvCount = SendCount[0];
   }

 if(MyRank != Root) {
   Source = Root;
   DataSize = 1;
   MPI_Recv(&RecvCount, DataSize, MPI_INT, Source, Source_tag, MPI_COMM_WORLD, &status);
	}

	/* Scatter Vector A and Vector B */

   Mybuffer_A = (float *)malloc(RecvCount * sizeof(float));
	MPI_Scatterv( Vector_A, SendCount, Displacement, MPI_FLOAT,Mybuffer_A, RecvCount, MPI_FLOAT, 
					 Root, MPI_COMM_WORLD);

   Mybuffer_B = (float *)malloc(RecvCount * sizeof(float));
	MPI_Scatterv( Vector_B, SendCount, Displacement, MPI_FLOAT,Mybuffer_B, RecvCount, MPI_FLOAT, 
					 Root, MPI_COMM_WORLD);

	/* Calculate partial sum */

	MyFinalVector = 0.0;
	for(index = 0 ; index < RecvCount ; index++) 
		 MyFinalVector += (Mybuffer_A[index] * Mybuffer_B[index]);

	/* Gather final answer on process 0 */

	DataSize = 1;
   MPI_Reduce(&MyFinalVector, &FinalAnswer, DataSize, MPI_FLOAT, MPI_SUM, Root, MPI_COMM_WORLD);

	if(MyRank == 0){
	   CheckResultVector = 0.0;
		for(index = 0 ; index < VectorSize ; index++) 
		 CheckResultVector += (Vector_A[index] * Vector_B[index]);
		if(fabs((double)(FinalAnswer-CheckResultVector)) > Epsilon){
		   printf("Error %d\n",index);
			ValidOutput = 0;
		}
		if(ValidOutput)
		  printf("FinalAnswer = %f\n", FinalAnswer);
        else
		  printf("Result may be wrong\n");
    
      free(Displacement);
      free(SendCount);
      free(Vector_A);
      free(Vector_B);
	 
   }
  
   free(Mybuffer_A);
   free(Mybuffer_B);

   MPI_Finalize();
}

File: mpi-1x-hypack-2013/matrix-vector-matrix-comp-codes/f77-lang-matvect/vv_mult_blkstp_nonunf.f
c
c********************************************************************
c
c	C-DAC Tech Workshop : hyPACK-2013
c                October 15-18, 2013
c
c  Example 5.2         : vv_mult_blkstp_nonunf.f
c
c  Objective           : Vector_Vector Multiplication 
c                        (Using blockstriped Non-uniform Data Partitioning) 
c
c  Input               : Read files vdata1.inp for Vector_A and 
c                        vdata2.inp Vector_B
c
c  Output              : Process with Rank 0 prints the Vector-Vector 
c                        Multiplication results 
c
c  Necessary Condition : Number of Processes should be
c                        less than or equal to 8
c
c   Created             : August-2013
c
c   E-mail              : hpcfte@cdac.in     
c
c********************************************************************
c
	 program main

	 include 'mpif.h'

         integer DATA_SIZE, PROC_SIZE
         real Epsilon

	 parameter (Epsilon = 1.0E-10, DATA_SIZE = 16)
	 parameter (PROC_SIZE = 8)

  	 integer   Numprocs, MyRank
  	 integer   VectorSize
  	 integer   VectorSize_A, VectorSize_B 
  	 integer   index, iproc
  	 integer   Root, DataSize
	 integer   Destination, Destination_tag
	 integer   Source, Source_tag
  	 integer   Distribute_Cols, Remaining_Cols 
     	 integer   Displacement(PROC_SIZE)
     	 integer   RecvCount, SendCount(PROC_SIZE)

  	 real    Mybuffer_A(DATA_SIZE), Mybuffer_B(DATA_SIZE)  
	 real	   MyFinalVector, FinalAnswer
  	 real 	   Vector_A(DATA_SIZE), Vector_B(DATA_SIZE)
	 integer   status(MPI_STATUS_SIZE)

C   ........MPI Initialisation .......

   	 call MPI_INIT(ierror) 
   	 call MPI_COMM_RANK(MPI_COMM_WORLD, MyRank, ierror)
   	 call MPI_COMM_SIZE(MPI_COMM_WORLD, Numprocs, ierror)

	 Root = 0

   	 if(MyRank .eq. Root) then 
 	  open(unit=12, file = './data/vdata1.inp')
	  read(12,*) VectorSize_A
	  read(12,*) (Vector_A(i), i=1,VectorSize_A)

 	  open(unit=12, file = './data/vdata2.inp')
	  read(12,*) VectorSize_B
	  read(12,*) (Vector_B(i), i=1,VectorSize_B)
	 endif

C     .......Allocate memory and read data for vector A .....

      call MPI_BCAST(VectorSize_A, 1, MPI_INTEGER, Root, 
     $		 MPI_COMM_WORLD, ierror)

      call MPI_BCAST(VectorSize_B, 1, MPI_INTEGER, Root, 
     $		 MPI_COMM_WORLD, ierror)

       if(VectorSize_A .ne. VectorSize_B) then
	  if(MyRank .eq. Root) then
          print *,"Two Vectors can not be multiplied - exit"
	  endif
	  goto 100
       else
        VectorSize = VectorSize_A
	endif
	 
      if(VectorSize .lt. Numprocs) then
        if(MyRank .eq. Root) then
          print*,"VectorSize should be more than No of Processors"
        endif
        goto 100
      endif

C	 ....... Initial arrangement for Scatterv operation ....

        if (MyRank .eq. Root) then
        	Distribute_Cols = VectorSize/Numprocs
		Remaining_Cols  = mod(VectorSize, Numprocs)
		do iproc = 1, Numprocs
		  SendCount(iproc) = Distribute_Cols
		enddo

		do iproc=Remaining_Cols,1,-1
		  SendCount(iproc) = SendCount(iproc) + 1
		enddo

		Displacement(1) = 1
		do iproc = 1, Numprocs
	   	  disp_value = 0
	      	  do jproc = 1, iproc-1
	           disp_value = disp_value + SendCount(jproc)
		  enddo
		  Displacement(iproc) = disp_value
		enddo

	endif

C	 .......Send RecvCount to each process ........ 

		 Destination_tag = 0
		 Source_tag = 0
	 if(MyRank .eq. Root) then
		do iproc = 1, Numprocs-1
   		Destination = iproc
	   	DataSize = 1
	   	RecvCount = SendCount(iproc+1)
	 call MPI_SEND(RecvCount,DataSize,MPI_INTEGER,Destination, 
     $	      Destination_tag, MPI_COMM_WORLD, ierror)
		enddo
      		RecvCount = SendCount(1)
	 else
	 	Source = Root
	 	DataSize = 1
	 call MPI_RECV(RecvCount, DataSize, MPI_INTEGER, Source, 
     $        Source_tag, MPI_COMM_WORLD, status, ierror)
	 endif

C   ......... Scatter Vector A and Vector B .........

      call MPI_SCATTERV(Vector_A, SendCount, Displacement, MPI_REAL, 
     $     Mybuffer_A,RecvCount,MPI_REAL,Root,MPI_COMM_WORLD,ierror)

      call MPI_SCATTERV(Vector_B, SendCount, Displacement, MPI_REAL, 
     $	   Mybuffer_B,RecvCount,MPI_REAL,Root,MPI_COMM_WORLD,ierror)

C	  ........ Calculate partial sum ........ 
	 MyFinalVector = 0.0
	 do index = 1, RecvCount
           MyFinalVector = MyFinalVector + 
     $	                   Mybuffer_A(index)*Mybuffer_B(index)
	 enddo

C	 .... Collective computation : Final answer on process 0 ..
        call MPI_REDUCE(MyFinalVector, FinalAnswer, 1, MPI_REAL, 
     $	     MPI_SUM, Root, MPI_COMM_WORLD, ierror)

	 if(MyRank .eq. Root) then
	  print*,'FinalAnswer', FinalAnswer
	 endif

 100	 call MPI_FINALIZE(ierror)
	 stop
	 end



